modality-probe-cli/src/export/mod.rs

```diff
@@ -0,0 +1,340 @@
+        (GraphType::Cyclic, true) => println!("{}", graph.graph.into_topology().to_dot(&cfg)?),
+    }
+
+    Ok(())
+}
+
+struct Cfg {
+    pub probes: HashMap<u32, ProbeMeta>,
+    pub events: HashMap<(Uuid, u32), EventMeta>,
+    pub probes_to_components: HashMap<u32, Uuid>,
+}
+
+fn assemble_components(comp_dirs: &mut Vec<PathBuf>) -> Result<Cfg, ExportError> {
+    let mut probes = HashMap::new();
+    let mut probes_to_components = HashMap::new();
+    let mut events = HashMap::new();
+
+    let mut event_files = Vec::new();
+    let mut probe_files = Vec::new();
+    for dir in comp_dirs.iter_mut() {
+        dir.push("probes.csv");
+        probe_files.push(dir.clone());
+        dir.pop();
+        dir.push("events.csv");
+        event_files.push(dir.clone());
+    }
+    for pf in probe_files.iter_mut() {
+        let mut pr_rdr = csv::Reader::from_path(pf.clone()).map_err(|e| {
+            ExportError(format!(
+                "failed to open the probes file at {}: {}",
+                pf.display(),
+                e
+            ))
+        })?;
+        for res in pr_rdr.deserialize() {
+            let p: ProbeMeta =
+                res.map_err(|e| ExportError(format!("failed to deserialize a probe row: {}", e)))?;
+            pf.pop();
+            pf.push("Component.toml");
+            let comp = Component::from_toml(&pf);
+            probes.insert(p.id, p.clone());
+            probes_to_components.insert(p.id, comp.id.0);
+        }
+    }
+    for ef in event_files.iter_mut() {
+        let mut ev_rdr = csv::Reader::from_path(ef.clone()).map_err(|e| {
+            ExportError(format!(
+                "failed to open the events file at {}: {}",
+                ef.display(),
+                e
+            ))
+        })?;
+        for res in ev_rdr.deserialize() {
+            let e: EventMeta =
+                res.map_err(|e| ExportError(format!("failed to deserialize a event row: {}", e)))?;
+            ef.pop();
+            ef.push("Component.toml");
+            let comp = Component::from_toml(&ef);
+            events.insert((comp.id.0, e.id), e.clone());
+        }
+    }
+
+    add_internal_events(&mut events)?;
+
+    Ok(Cfg {
+        events,
+        probes,
+        probes_to_components,
+    })
+}
+
+fn add_internal_events(events: &mut HashMap<(Uuid, u32), EventMeta>) -> Result<(), ExportError> {
+    let nil_uuid = Uuid::nil();
+    for ie in Events::internal_events() {
+        let ev = EventMeta {
+            component_id: nil_uuid,
+            id: ie.id.0,
+            name: ie.name,
+            type_hint: if ie.type_hint.is_empty() {
+                None
+            } else {
+                Some(ie.type_hint)
+            },
+            tags: ie.tags,
+            description: ie.description,
+            file: ie.file,
+            line: ie.line,
+        };
+        events.insert((nil_uuid, ie.id.0), ev.clone());
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
A high-level test here (or an external integration test) demonstrating that each combination of the CLI options can produce valid, non-empty graphviz-compatible output would be most awesome.

Reply:
Totally agree. There is not enough testing here at all.